examples/bin/start-druid-main

```diff
@@ -0,0 +1,436 @@
+
+def check_memory_constraint(total_memory, service_list):
+    # 80% of total memory >= sum of lower bound service memory should be
+    lower_bound_memory = 0
+
+    for service in service_list:
+        lower_bound_memory += SERVICE_MEMORY_LOWER_BOUND.get(service)
+
+    required_memory = int(lower_bound_memory / 0.8)
+
+    if total_memory < required_memory:
+        error_and_exit(f'Minimum memory required for starting services is {required_memory}m')
+
+    if total_memory >= 2 * lower_bound_memory:
+        return int(total_memory / 2)
+    else:
+        return lower_bound_memory
+
+def build_mm_task_javaopts_array(memory_type):
+    task_memory = f'{MM_TASK_JAVAOPTS_PROP}=['
+
+    MEM_ARRAY = MM_TASK_MEM_MAP.get(memory_type)
+
+    javaopts_list = MM_TASK_JAVAOPTS_ARRAY + MEM_ARRAY
+
+    for item in javaopts_list:
+        task_memory += f'\"{item}\";'
+
+    task_memory = task_memory[:-1]
+    task_memory += ']'
+    return task_memory
+
+def build_memory_config_string(service, allocated_memory):
+    if service == MM_TASK:
+        if allocated_memory >= 2048:
+            task_count = int(allocated_memory / 2048)
+            memory_type = MM_TASK_MEMORY_TYPE_HIGH
+            task_memory = 2048
+        else:
+            task_count = int(allocated_memory / 512)
+            memory_type = MM_TASK_MEMORY_TYPE_LOW
+            task_memory = 512
+        task_count = min(task_count, multiprocessing.cpu_count())
+
+        javaopts_array = build_mm_task_javaopts_array(memory_type)
+        return [f'{MM_TASK_WORKER_CAPACITY_PROP}={task_count}', javaopts_array], task_memory * task_count
+    else:
+        heap_memory = SERVICE_MEMORY_HEAP_PERCENTAGE.get(service) * allocated_memory
+        direct_memory = int(allocated_memory - heap_memory)
+        heap_memory = int(heap_memory)
+
+        if direct_memory == 0:
+            return f'-Xms{heap_memory}m -Xmx{heap_memory}m', allocated_memory
+
+        return f'-Xms{heap_memory}m -Xmx{heap_memory}m -XX:MaxDirectMemorySize={direct_memory}m', allocated_memory
+
+def distribute_memory_over_services(services, total_memory):
+    service_memory_config = {}
+
+    memory_weight_sum = 0
+
+    service_list = services.copy()
+    if MM in services:
+        service_list.append(MM_TASK)
+
+    for service in service_list:
+        memory_weight_sum += SERVICE_MEMORY_DISTRIBUTION_WEIGHT.get(service)
+
+    multiplier = total_memory / memory_weight_sum
+
+    lower_bound_memory_allocation = 0
+    allocated_services = set()
+
+    for service in service_list:
+        allocated_memory = SERVICE_MEMORY_DISTRIBUTION_WEIGHT.get(service) * multiplier
+        if service in SERVICE_MEMORY_LOWER_BOUND and allocated_memory < SERVICE_MEMORY_LOWER_BOUND.get(service):
+            allocated_memory = SERVICE_MEMORY_LOWER_BOUND.get(service)
+            service_memory_config[service], allocated_memory = build_memory_config_string(service, allocated_memory)
```

Comment:
Why is allocated_memory being reassigned here?

Reply:
Memory allocated to task is multiple of either 512MB or 2GB.
So the reassignment is basically this computation:
`allocated_memory = allocated_memory - task_count * task_memory`.

Reply:
Makes sense.

Style suggestion:
If this is going to happen only for tasks, it would be better to have a separate logic for tasks in this function itself rather than expecting the common function `build_memory_config_string` to do the right thing. You can maybe break up the `build_memory_config_string` into two methods.

For both cases, it would be cleaner to have the memory (and task count) computed and returned by that function and construct the config string here itself. (This is to avoid having a name such as `build_memory_config_string` which is actually also updating the allocated memory and other params)